cpp/src/arrow/compute/kernels/random.cc

```diff
@@ -0,0 +1,85 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <memory>
+#include <random>
+
+#include "arrow/builder.h"
+#include "arrow/compute/api_scalar.h"
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/kernels/common.h"
+#include "arrow/compute/registry.h"
+#include "arrow/util/pcg_random.h"
+
+namespace arrow {
+namespace compute {
+namespace internal {
+namespace {
+
+// Generates a random floating point number in range [0, 1).
+double generate_uniform(random::pcg64& rng) {
+  return (rng() >> 11) * (1.0 / 9007199254740992.0);
+}
+
+using RandomState = OptionsWrapper<RandomOptions>;
+
+Status ExecRandom(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+  static thread_local std::random_device rd;
+  static thread_local random::pcg64 gen;
```

Comment:
Normal stack variable is okay, without `static thread_local` IMO.
Shall we use `random::pcg64_fast` instead?

Reply:
Switched to `random::pcg64_fast`, seems to work.

Regarding `thread_local`, not sure. Isn't it possible that we'd call `random` from different threads at the same time? Both `std::random_device` and `random::pcg64` are unsafe to share (without locking) in that case, which is why I'm making them `thread_local`. But if there's a higher-level synchronization then I can leave `thread_local` out.

Reply:
I mean normal variable on **stack**, no `static`, no `thread_local`.

Reply:
Done. I was worried about re-initializing `std::random_device` over and over again, that can be slow, but given that we now return arrays should be fine.

Reply:
Regarding the cost of re-initializing `std::random_device`, I did a quick benchmark (on apple m1).
- **test1** re-init `std::random_device` in each iteration, it finishes in 0.6s.
- **test2** init `std::random_dedvice` only once, if finishes in 0.07s.

The overhead of initialize `std::random_device` (probably due to syscalls like open("/dev/random")) is about 5us (600000/123456) roughly.

I think it's not a problem in practice. But a benchmark will be more convincing.
@asuhan , if you don't mind, I'd like to push extra commits to this PR to add a benchmark (author of this PR will still be you).

**test1** (finishes in 0.6s)
```
#include <random>
int main() {
  int s = 0;
  for (int i = 0; i < 123456; ++i) {
    std::random_device rd;    // re-init in each loop
    s += rd();
  }
  return s;
}
```

**test2** (finishes in 0.07s)
```
#include <random>
int main() {
  int s = 0;
  std::random_device rd;    // init only once
  for (int i = 0; i < 123456; ++i) {
    s += rd();
  }
  return s;
}```

Reply:
+1 for adding a benchmark. I'm curious about the overhead on different OSes.

Reply:
Sure, go ahead with adding some benchmark commits.

Reply:
On linux/x86 there's no big gap between above two tests (about 30%).
From `dtruss/strace`, macos/m1 is reading /dev/urandom (`open` when defining random_device, `read` when calling operator()), on linux/x86 there's no syscall at all, guess it's using intel rdrand feature[1]. The difference is probably due to cpu architecture, not os/compiler.

[1] https://en.wikipedia.org/wiki/RDRAND